### PR TITLE
feat(ProtoRev): Supporting Two Pool Routes

### PR DESCRIPTION
**Pull request by davidterpay**
## What is the purpose of the change

This PR updates the route building logic of `ProtoRev` to be able to construct two pool arbitrage routes. This is particularly useful when there are multiple pools of that consist of the same assets and are imbalanced. As before, routes will only be constructed using the base denoms. 

## Testing and Verifying

  - *Added unit test that test the construction of two pool routes under various conditions*
  - *Updated existing test cases that should also now create two pool routes*

## Documentation and Release Note

This PR is a part of the open issue for `ProtoRev` updates for V17 (https://github.com/osmosis-labs/osmosis/issues/5858)

### CODE REVIEW

File: x/protorev/keeper/routes.go [hunk 1]
```diff
@@ -149,6 +153,51 @@ func (k Keeper) BuildHighestLiquidityRoute(ctx sdk.Context, swapDenom types.Base
 	}, nil
 }
 
+// BuildTwoPoolRoute will attempt to create a two pool route that will rebalance pools that are paired with the base denom.
+// This is useful for pools that contain the same assets but are imbalanced.
+func (k Keeper) BuildTwoPoolRoute(ctx sdk.Context, baseDenom types.BaseDenom, tokenIn, tokenOut string, poolId uint64) (RouteMetaData, error) {
+	if baseDenom.Denom != tokenOut {
```
Comment (NotJeremyLiu) [inline x/protorev/keeper/routes.go:159]: Synced in person about supporting both directions

File: x/protorev/keeper/keeper_test.go [hunk 2]
```diff
@@ -902,6 +903,91 @@ func (s *KeeperTestSuite) setUpPools() {
 	// Pool 50
 	s.PrepareCosmWasmPool()
 
+	// Create a duplicate pool for testing
```
Comment (NotJeremyLiu) [inline x/protorev/keeper/keeper_test.go:906]: Can we add in the comment what test these pools are for

**Review by NotJeremyLiu — APPROVED** [review]
lgtm

File: x/protorev/keeper/routes_test.go [hunk 3]
```diff
@@ -225,6 +246,99 @@ func (s *KeeperTestSuite) TestBuildHighestLiquidityRoute() {
...
+			tokenOut: "stake",
+			poolId:   53,
+			expectedRoute: []TestRoute{
+				{PoolId: 54, InputDenom: types.OsmosisDenomination, OutputDenom: "stake"},
+				{PoolId: 53, InputDenom: "stake", OutputDenom: types.OsmosisDenomination},
+			},
+			hasRoute: true,
+		},
+		{
+			description: "two pool route where swap is on the highest liquidity pool",
```
Comment (stackman27) [inline x/protorev/keeper/routes_test.go:291]: i'm trying to understand this: if SwapDenom == tokenOutDenom & huge swap occours in highest liquidity pool, isn't based denom being cheaper so you **would** create a route to balance it?

**Review by stackman27 — COMMENTED** [review]
just have a question to understand this logic more before approving

File: x/protorev/keeper/routes_test.go (hunk 3 above)
Comment (davidterpay) [reply x/protorev/keeper/routes_test.go:291]: just to make sure i understand the scenario: 

if `BaseDenom` == `tokenOutDenom`, base denom becomes more expensive as the token supply of `baseDenom` in that pool decreases. i.e. if x == tokenInDenom and y == tokenOutDenom, then x + deltaX / y - deltaY > x / y. in this case, tokenOut is y which means it is more expensive and hence we can get more of X on the current pool than the other pool. We should then find the highest liquidity pool with the same assets and route through that. if the swap occurred on the highest liquidity pool, then we cannot make a route.

**Comment by czarcas7ic** [comment]
@NotJeremyLiu @davidterpay getting the same protorev e2e failure after two runs. Any way you can take a look into this?

**Comment by davidterpay** [comment]
> @NotJeremyLiu @davidterpay getting the same protorev e2e failure after two runs. Any way you can take a look into this?

It looks like its failing a CL test swap. I'll see if I can replicate the same issue and resolve it locally. Will ping once I find the issue (likely tmw morning EST).

**Comment by ValarDragon** [comment]
devbot merge base

**Comment by davidterpay** [comment]
We think the issue may be arising because of the two pool routing as the swap in the CL test does create some arbitrage that the module is confirmed to capture. This might be messing with the calculations done in the test case.

There are a few approaches for this:
1. We can disable protorev for the test case (but im unsure if this will affect the other protorev test since it is executed in parallel).
2. We can determine if the calculations for sqrt amount can be updated in accordance with any liquidity changes that protorev introduces.
3. We can switch out the denoms used in the test case for other denoms so that we do not have to disable protorev.

**Comment by davidterpay** [comment]
```
// Change the parameter to disable protorev
err = chainABNode.ParamChangeProposal("protorev", string(protorevtypes.ParamStoreKeyEnableModule), []byte("false"), chainAB)
s.Require().NoError(err)
```

Added this for the time being. Should be passing E2E now.

**Comment by czarcas7ic** [comment]
Do you think we can instead remove the route from protorev in the protorev test itself? It doesn't seem scalable if we have another test that relies on calculating the output of the pools to check if protorev is disabled. Also, since everything is concurrent, there could be a scenario in which
1. Protorev test starts
2. CL test starts as well
3. CL test diables protorev
4. Protorev test fails due to being disabled via the CL test

**Comment by davidterpay** [comment]
@czarcas7ic set the max pool points such that it will always pass on the ProtoRev test and will never execute a backrun in the CL test. Should always pass now even if the tests execute in parallel.

File: tests/e2e/configurer/chain/commands.go [hunk 4]
```diff
@@ -770,6 +777,7 @@ func (n *NodeConfig) ParamChangeProposal(subspace, key string, value []byte, cha
 		Deposit: "625000000uosmo",
 	}
 	proposalJson, err := json.Marshal(proposal)
+	fmt.Println("marshalling output", proposalJson, err)
```
Comment (czarcas7ic) [inline tests/e2e/configurer/chain/commands.go:780]:
```suggestion
```

**Review by czarcas7ic — APPROVED** [review]

File: tests/e2e/configurer/chain/commands.go (hunk 4 above)
Comment (davidterpay) [reply tests/e2e/configurer/chain/commands.go]: forgot to remove after testing. ty for removing this.

**Pull request merged**